### PR TITLE
[TargetLowering] Use Correct VT for Multi-out Asm

### PR DESCRIPTION
**Pull request by lenary**
This was overlooked in 7d940432c46be83b8fcb5dbefee439585fa820cd - when inline assembly has multiple outputs, they are returned as members of a struct, and the `getAsmOperandType` needs to be called for each member of struct. The difference between this and the single-output case is that in the latter, there isn't a struct wrapping the outputs.

I noticed this when trying to use the same mechanism in the RISC-V backend.

Committing two tests:
- One that shows a crash before this change, which is fixed by this change.
- One (commented out) that shows a different crash with tied inputs/outputs. This is commented as it is not fixed by this change and needs more work in target-independent inline asm handling code.

### CODE REVIEW

File: llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp [hunk 1]
```diff
@@ -5753,7 +5753,8 @@ TargetLowering::ParseConstraints(const DataLayout &DL,
       assert(!Call.getType()->isVoidTy() && "Bad inline asm!");
       if (auto *STy = dyn_cast<StructType>(Call.getType())) {
         OpInfo.ConstraintVT =
-            getSimpleValueType(DL, STy->getElementType(ResNo));
+            getAsmOperandValueType(DL, STy->getElementType(ResNo))
+                .getSimpleVT();
```
Comment (arsenm) [inline llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:5757]: What happens for nested structs? Are those supported?

**Review by arsenm — APPROVED** [review]

**Review by labrinea — COMMENTED** [review]
I haven't touched this code in ages, but looks sensible.

File: llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp (hunk 1 above)
Comment (lenary) [reply llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:5757]: I don't believe the lowering from an inline assembly statement to a `call ... asm` will create a nested struct. It either returns the single type corresponding to the output (if there's one output), or a struct with a member per output (if there's multiple outputs). I think that inline asm inputs/outputs cannot themselves be structs, but I'm not 100% certain of that. This code would suggest not.

File: llvm/test/CodeGen/AArch64/ls64-inline-asm.ll [hunk 2]
```diff
@@ -103,3 +103,37 @@ entry:
   call void asm sideeffect "st64b $0,[$1]", "r,r,~{memory}"(i512 %s.sroa.0.0.insert.insert, ptr %addr)
   ret void
 }
+
```
[the remark this answers is not in the document]
Comment (lenary) [reply llvm/test/CodeGen/AArch64/ls64-inline-asm.ll:104]: Done

File: llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp (hunk 1 above)
Comment (arsenm) [reply llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:5757]: I just tried it and it seems to hit unreachable "Unknown type!"; we should do better here and report a proper error

Comment (lenary) [reply llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:5757]: Better errors for IR that frontends don't produce seems like a different issue to fixing an actual crash in the AArch64 backend that frontends can hit, so should be in a different PR IMO.

Comment (arsenm) [reply llvm/lib/CodeGen/SelectionDAG/TargetLowering.cpp:5757]: Yes, separate fix

**Pull request merged**